base/abstractarray.jl

```diff
@@ -152,39 +157,19 @@ checkbounds(A::AbstractArray, I...) = (@_inline_meta; _internal_checkbounds(A, I
 # _checkbounds previously that meant something different.
 _internal_checkbounds(A::AbstractArray, I::AbstractArray{Bool}) = size(A) == size(I) || throw_boundserror(A, I)
 _internal_checkbounds(A::AbstractArray, I::AbstractVector{Bool}) = length(A) == length(I) || throw_boundserror(A, I)
-_internal_checkbounds(A::AbstractArray, I) = (@_inline_meta; checkbounds(Bool, length(A), I) || throw_boundserror(A, I))
-function _internal_checkbounds(A::AbstractMatrix, I, J)
+function _internal_checkbounds(A::AbstractArray, I...)
     @_inline_meta
-    (checkbounds(Bool, size(A,1), I) && checkbounds(Bool, size(A,2), J)) ||
-        throw_boundserror(A, (I, J))
-end
-function _internal_checkbounds(A::AbstractArray, I, J)
-    @_inline_meta
-    (checkbounds(Bool, size(A,1), I) && checkbounds(Bool, trailingsize(A,Val{2}), J)) ||
-        throw_boundserror(A, (I, J))
-end
-@generated function _internal_checkbounds(A::AbstractArray, I...)
-    meta = Expr(:meta, :inline)
-    N = length(I)
-    Isplat = [:(I[$d]) for d=1:N]
-    error = :(throw_boundserror(A, tuple($(Isplat...))))
-    args = Expr[:(checkbounds(Bool, size(A,$dim), I[$dim]) || $error) for dim in 1:N-1]
-    push!(args, :(checkbounds(Bool, trailingsize(A,Val{$N}), I[$N]) || $error))
-    Expr(:block, meta, args...)
+    checkbounds(Bool, size(A), I...) || throw_boundserror(A, I)
 end
 
 ## Bounds-checking without errors ##
-function checkbounds(::Type{Bool}, sz::Dims, I...)
-    n = length(I)
-    for dim = 1:(n-1)
-        checkbounds(Bool, sz[dim], I[dim]) || return false
-    end
-    s = sz[n]
-    for i = n+1:length(sz)
-        s *= sz[i]
-    end
-    checkbounds(Bool, s, I[n])
+function checkbounds{N}(::Type{Bool}, sz::NTuple{N,Integer}, I1, I...)
+    @_inline_meta
+    checkbounds(Bool, sz[1], I1) || return false
+    checkbounds(Bool, tail(sz), I...)
```

Comment:
Gist: https://gist.github.com/timholy/12050e58bc2ad5c38987

Master:
```
julia> include("perf_checkbounds.jl")
  0.101018 seconds (6 allocations: 192 bytes)
  0.199976 seconds (3.00 M allocations: 381.470 MB, 7.73% gc time)
  1.756785 seconds (100.00 M allocations: 2.980 GB, 7.38% gc time)
  2.577147 seconds (20.00 M allocations: 610.352 MB, 1.08% gc time)
10000000
```

This PR:
```
julia> include("perf_checkbounds.jl")
  0.101331 seconds (6 allocations: 192 bytes)
  0.250744 seconds (4.00 M allocations: 427.246 MB, 17.92% gc time)
  0.605173 seconds (7 allocations: 224 bytes)
  0.144111 seconds (9 allocations: 368 bytes)
```
Maybe a small regression in one case, but a win in another.

I will run the whole Array suite, however.